specifications/unified-provider.org

```diff
@@ -0,0 +1,84 @@
+#+TITLE: Unified Provider
+#+AUTHOR: Chris Patterson <chris.patterson@canonical.com>
+#+DATE: [2020-09-11]
+
+* Scope:
+
+Snapcraft currently implements build providers for LXD and Multipass.  Host
+builds are driven with independent logic, and Launchpad is indendently
+implemented for remote-build.  With the re-introduction of remote LXD support,
+it is desirable to refactor all providers into a consistent implementation that
+has varied support for Snapcraft capabilities.
+
+This document defines an approach to refactor build provider
+=snapcraft.internal.build_providers._base_provider.Provider= such that it would
+support all current build providers (Host, Launchpad, LXD, Managed-Host,
+Multipass) and potential future providers (Remote LXD, etc.).
+
+* Provider Refactoring
+
+** Current Public Interfaces
+- ensure_provider()
+- setup_provider()
+- mount_project()
+- expose_prime()
+- execute_step()
+- pack_project()
+- clean(part_names)
+- clean_project()
+- shell()
+
+** Proposed Public Interfaces:
+
+|--------------------------------------------+---------------------------------------------------------------------------+-------------------------------------------------------|
```

Comment:
How about making this class like (ABC) where methods are methods, descriptions and raises docstrings

Reply:
But org tables are so much fun... i'll do that.